Add vertex light support

magenta404 opened the pull request:
This will help address #357
There is also pixel light support with #383

Experimental support for vertex lights. Vertex lights are point lights which are set to _Not Important_ or in some cases _Automatic_.

![Night](https://user-images.githubusercontent.com/5249806/71505496-7c7c0580-28d1-11ea-9f5a-6c4e7094686f.jpg)
Night Time

## About

### Attenuation
The attenuation formula was taken from [here](https://forum.unity.com/threads/point-light-in-v-f-shader.499717). The `UNITY_LIGHT_ATTENUATION` doesn't work for vertex lights. The `Shade4PointLights`, as discussed in the thread, didn't give good results.

### Shader Parameters
I've added some shader parameters so others for flexibility for the time being. Not sure if they will be merged. They are mostly there to separate the parameters for the primary direction light.

### Notes
`ShadeVertexLightsFull` cannot be used since it is for _Legacy Vertex Lit_.

## Issues

### Tiling Artefacts
![Tiling](https://user-images.githubusercontent.com/5249806/71506195-088f2c80-28d4-11ea-9617-08927ec8fcbd.jpg)

Tiling artefacts occur depending on the attenuation, lights overlapping and camera orientation. The frame debugger shows lighting values being filled. It could be the attenuation calculation or a limitation in vertex lighting. Further investigation required.

### Underwater
![Underwater](https://user-images.githubusercontent.com/5249806/71509401-404fa180-28df-11ea-8144-e15b4a79a8f5.jpg)

I have come across this issue when trying to integrate a weather asset with cloud shadows. It is much worse at night. Not including lights might be the best solution. Or at least having it behind a toggle.

## Missing

- Directional Light style reflections

magenta404 committed:
Add vertex lights support. Experimental

huwb commented:
Really solid work. Code looks all good to me.

I tried it out and seems to work well from initial experiments.

Could you push the tiling fail case? You could check in main.unity temporarily and we could revert it later, if that makes sense.

daleeidd commented:
Thanks. Done

huwb commented:
It may be because the light is set to Auto and unity is deciding to make it a pixel light. If i set the red light to be "Not Important" the issue goes away.. ?

daleeidd commented:
I thought that too. I am in the same test case now and I have to set them (or at least the red one) to _Important_ for the issue to go away.

![ABROKE](https://user-images.githubusercontent.com/5249806/71515352-efe43e00-28f6-11ea-9b8f-ec8eb3a39328.jpg)
![ANOTBROKE](https://user-images.githubusercontent.com/5249806/71515353-efe43e00-28f6-11ea-92ee-32b5017ffd79.jpg)

daleeidd commented:
I just went through the frame debugger and `VERTEXLIGHT_ON` was active for those tiles.

Metal is a bit strange with lighting. I will have to test on Windows tomorrow. Hopefully, I can get the same as yours.

daleeidd commented:
I get the same for Windows as for Mac. The issue is that the red colour is severely diminished:

Lights marked as _Important_
![A Important](https://user-images.githubusercontent.com/5249806/71538633-77c35a00-2982-11ea-8410-4cc76d48f255.png)

Lights marked as _Not Important_
![A Not Important](https://user-images.githubusercontent.com/5249806/71538640-83168580-2982-11ea-80a5-8f25d89f388c.png)

Also, if I change one of the green lights to a blue colour, the red colour is passed through correctly.

This could be a bug in Unity.

magenta404 committed:
Add comment about Shade4PointLights

magenta404 committed:
Multiply bubbles by lights

Multiplying gave bad results for ambient light in scattering. Works
better with bubbles than addition.

magenta404 committed:
Add/improve shader param comments

huwb commented:
Thanks for the above, and trying to wrap my head around things.

I think that Unity will move lights between vertex and pixel depending on number of lights, importance, maybe intensity(?). And there is a limit of 4 lights per vertex (and the directional light seems to impact the result for me - when i turn it off the result changes - so it's counted in the 4?).

If all that is true the story that my brain has constructed is that it is moving lights between vertex and pixel based on these factors, but this branch only has vertex, not pixel, so lights disappear. But I'm not 100% sure if that is consistent with what we're seeing / with the above?

huwb commented:
Worth noting as well that since verts can be displaced quite far horizontally, the render bounds are quite a lot larger than the tile appears and it may be overlapping more lights than it would seem

![image](https://user-images.githubusercontent.com/939901/71542330-08873f00-295d-11ea-9102-9ed0a4f9db7f.png)

daleeidd commented:
Thanks. Displacement tip is useful. After disabling attenuation:
![Displacement Lighting](https://user-images.githubusercontent.com/5249806/71542936-06ef6280-29c1-11ea-944d-faa441b4a04e.jpg)

I have gone over the following again:
https://docs.unity3d.com/Manual/RenderTech-ForwardRendering.html

> but this branch only has vertex, not pixel, so lights disappear

It is hard to say. The only thing I think I can do to confirm is to merge the pixel and vertex branch together, and max out the lights to see what happens. Not too difficult.

> But I'm not 100% sure if that is consistent with what we're seeing / with the above?

From what I have been experiencing, it hasn't been consistent with the manual. I didn't think _Important_ lights would even render without pixel light support in the shader. I thought that Unity was populating the shader data with pixel light calculations. I have also seen the issue with lower amounts of lights. Lastly, when comparing _Important_ and _Not Important_ in the Frame  Debugger, the light positions, world light and spherical harmonics are the same; the only thing changed is the light colours. I would think if the light disappears, then the light positions shouldn't include the same set of lights:

Important
![111 Bad](https://user-images.githubusercontent.com/5249806/71544174-f85c7780-29cf-11ea-9c37-135f1768ecce.jpg)
Not Important
![111 Good](https://user-images.githubusercontent.com/5249806/71544178-fd212b80-29cf-11ea-94b4-fcbdc31d2825.jpg)

I might have to make some reproducible cases with planes and a trimmed down shader. I'll let this sit in my mind for a bit.

daleeidd commented:
I will make a forum post or file a bug since the issue is unrelated to Crest.

daleeidd commented:
Finally got around to it: https://forum.unity.com/threads/vertex-lights-behaving-strangely.822300/

daleeidd commented:
Update from Unity (12/03/2020):
> We successfully reproduced this issue and have sent it for resolution with our developers.

daleeidd commented:
Removing draft status. It could probably use more work which I can look into another time.

Unity has given a conclusion:
> This particular case has been investigated thoroughly and we have decided, to not address this fix for the time being. Our developers found out that using stock URP shader works as expected, so you could use that as a workaround.

We could implement it here and then recommend URP if anyone wants nice lighting.

magenta404 committed:
Merge branch 'master' into feature/vertex-lights

  # Conflicts:
  #	crest/Assets/Crest/Crest/Shaders/Ocean.shader
  #	crest/Assets/Crest/Crest/Shaders/OceanEmission.hlsl
  #	crest/Assets/Crest/Crest/Shaders/OceanFoam.hlsl
  #	crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterCurtain.shader

magenta404 committed:
Re index texcoords

magenta404 committed:
Apply depth fog to SSS vertex light colour

magenta404 committed:
Housekeeping
